### PR TITLE
[BREAKING][ops, model] feat: GPU-optimal ops defaults + strict NPU validation

### PR DESCRIPTION
**Pull request by TimYangst**
### What does this PR do?

> Flip `OpsImplementationConfig` per-op defaults from `"eager"` to GPU-optimal
> (Liger / Triton / fused_triton), tighten the NPU validator and per-model
> dispatch, and remove the silent all-eager fallback in
> `build_foundation_model` so misconfigurations raise loudly instead of
> silently downgrading.
>
> Default changes (GPU):
>
> | Field | Old | New |
> |---|---|---|
> | `attn_implementation` | `flash_attention_2` | unchanged |
> | `moe_implementation` | `eager` | `fused_triton` |
> | `cross_entropy_loss_implementation` | `chunk_loss` | `liger_kernel` |
> | `rms_norm_implementation` | `eager` | `liger_kernel` |
> | `swiglu_mlp_implementation` | `eager` | `liger_kernel` |
> | `rotary_pos_emb_implementation` | `eager` | `liger_kernel` |
> | `load_balancing_loss_implementation` | `eager` | `triton` |
> | `rms_norm_gated_implementation` | `eager` | unchanged (Qwen3.5; FLA, GPU-only) |
> | `causal_conv1d_implementation` | `eager` | unchanged (Qwen3.5; FLA, GPU-only) |
> | `chunk_gated_delta_rule_implementation` | `eager` | unchanged (Qwen3.5; FLA, GPU-only) |
>
> Three behaviour changes flow from this:
>
> 1. **NPU users must pin every per-op field explicitly** — to an NPU-supported
>    value (`npu` / `chunk_loss` / `fused_npu` / `triton`) or `eager` when the
>    op has no NPU backend. The validator raises at config-parse time with a
>    per-op allow-list. The Qwen3.5 GatedDeltaNet trio (`rms_norm_gated` /
>    `causal_conv1d` / `chunk_gated_delta_rule`, added in #714) is validated
>    at OpSlot-bind time instead — these slots only exist in Qwen3.5's
>    patched module, so config-parse validation would force every NPU user
>    to override them even when training non-Qwen3.5 models. The kernel's
>    `HardwareRequirement` is the single source of truth.
>
> 2. **Per-model dispatch is strict-only.** `apply_per_model_patches` raises
>    on every missing backend except `eager`. Models with structurally
>    incompatible kernels (Wan `rope_apply`, Qwen2-VL multimodal RoPE on NPU)
>    opt out via `extra_backends[name] = None`, which is now an
>    *explicit-raise marker* with a `"explicitly disabled for {model}"`
>    message instead of a silent skip.
>
> 3. **`build_foundation_model` requires explicit ops.** Callers must pass
>    `ops_implementation=...` or pre-install a singleton via
>    `apply_ops_config(...)`; passing neither raises `ValueError`. The prior
>    silent all-eager fallback masked bugs like `train_wan` / `seed_omni`
>    passing only `attn_implementation` and having it silently overridden.
>
> Builds on #678 (registration-based kernel framework), #708 (chunk_loss
> reshape fix), #711 (chunked log-probs), #714 (Qwen3.5 GatedDeltaNet OpSlot
> dispatch), and #720 (temperature/entropy in log-probs forward).

### API and Usage Example

> **GPU**: existing YAMLs that don't override the per-op fields now pick up
> the GPU-optimal kernels automatically. The `gpu` extra ships
> `liger-kernel` + `triton`, so no setup change is needed.
>
> **NPU**: pin every per-op field. Minimal NPU-friendly YAML:
>
> ```yaml
> model:
>   ops_implementation:
>     attn_implementation: flash_attention_2
>     moe_implementation: fused_npu
>     cross_entropy_loss_implementation: chunk_loss
>     rms_norm_implementation: npu
>     rotary_pos_emb_implementation: npu
>     swiglu_mlp_implementation: eager           # no NPU backend
>     load_balancing_loss_implementation: eager  # triton-ascend not exposed as `triton`
>     # Qwen3.5 GatedDeltaNet trio — only meaningful for Qwen3.5 MoE; pin to
>     # eager on NPU (no NPU kernel) and set train.dyn_bsz=False (eager path
>     # only supports non-varlen training):
>     rms_norm_gated_implementation: eager
>     causal_conv1d_implementation: eager
>     chunk_gated_delta_rule_implementation: eager
> ```
>
> **Models with structurally-incompatible kernels** (Wan `rope_apply`,
> Qwen2-VL multimodal RoPE on NPU) ship pinned. Wan's three YAMLs include
> `rotary_pos_emb_implementation: eager`.

### Design & Code Changes

> **Defaults & validation** — `veomni/arguments/arguments_types.py`,
> `veomni/ops/kernels/{rms_norm,rotary,swiglu,load_balancing_loss}/__init__.py`
>
> - Flip per-op dataclass defaults; sync each `OpSpec.default` to match.
> - Rewrite `_validate_implementations`:
>   1. Hard-assert at parse time that every registered op (`list_ops()`) has
>      an entry in `_NPU_ALLOWED` — catches future op additions that miss
>      the map.
>   2. Check NPU compatibility against the explicit `_NPU_ALLOWED`
>      allow-list with a per-op error message.
>   3. Gate `load_balancing_loss=triton` on `is_package_available("triton")`.
> - Per-package availability for `liger_kernel` / `torch_npu` / `triton` is
>   checked at the resolution sites (`apply_per_model_patches` /
>   `apply_global_ops`), not duplicated in the validator.
>
> *Trade-off — hardcoded allow-list*: `_NPU_ALLOWED` is hand-curated rather
> than derived from `BackendSpec.requires`, because the name `"triton"` is
> reused with different hardware semantics (NPU `triton-ascend` for
> load-balancing loss vs. CUDA-only mainline Triton for DeepSeek-V3
> RMSNorm/RoPE). A `requires`-based inference would conflate them.
> Third-party NPU backends still extend cleanly via per-model
> `extra_backends` in `device_patch.py`.
>
> **Strict ops contract** — `veomni/models/auto.py`
>
> - Remove the silent all-eager fallback. `build_foundation_model` raises
>   `ValueError` when neither `ops_implementation` is passed nor a prior
>   `apply_ops_config(...)` has installed a singleton. Trainers always pass
>   `args.model.ops_implementation`. Callers that previously relied on the
>   fallback now construct an explicit `OpsImplementationConfig`:
>   - `tasks/infer/{infer_text,infer_qwen2_vl,infer_omni_model}.py` — pin
>     every per-op field to `eager`, with `flash_attention_2` for attn when
>     the package is importable so they keep working on CPU/minimal-deps
>     hosts;
>   - `tasks/deprecated_task/train_wan.py` — pass full
>     `args.model.ops_implementation` instead of just `attn_implementation`;
>   - `veomni/models/seed_omni/auto.py::build_omni_model` — accept and
>     forward an `ops_implementation` kwarg; `tasks/omni/train_omni_model.py`
>     passes `args.model.ops_implementation`.
> - `attn_implementation` resolution: caller passes `ops_implementation` →
>   use `ops.attn_implementation` (kwarg ignored); caller pre-installed the
>   singleton via `apply_ops_config` → use the singleton's
>   `attn_implementation` unless an explicit kwarg was passed (the explicit
>   value wins); neither passed and no singleton → raise.
> - `OpsImplementationConfig.all_eager()` is removed (it was an escape hatch
>   masquerading as public API). Tests that need a hardware-agnostic config
>   use `tests/tools/training_utils.py::make_eager_ops_config(**overrides)`;
>   NPU tests use `make_npu_ops_config(model_name=...)` for the recommended
>   NPU backend per op (consolidates the `_NPU_OPS_DEFAULTS` /
>   `_NPU_PER_MODEL_OVERRIDES` table that already powered
>   `resolve_ops_overrides`).
>
> **Per-model device patches**
>
> - `qwen3_vl/device_patch.py` — drop historical `liger_kernel: None`
>   opt-outs (verified standard-shape RMSNorm + RoPE).
> - `wan/device_patch.py` — keep `rotary_pos_emb: liger_kernel: None` as the
>   explicit-raise marker; **move Wan's Triton RoPE from the `_custom_wan`
>   callback into `extra_backends["rotary_pos_emb"]["triton"]`** so it
>   survives the strict-raise gate (a callback can't run after the gate
>   rejects unknown backends).
> - `qwen2_vl/device_patch.py` — keep `rotary_pos_emb: npu: None` as the
>   explicit-raise marker for the multimodal RoPE NPU mismatch.
> - `qwen3_5_moe/__init__.py` — Qwen3.5 GatedDeltaNet's NPU rejection runs
>   solely through `KERNEL_REGISTRY.resolve`'s `HardwareRequirement` check
>   (single source of truth). The varlen `dyn_bsz=False` caveat is
>   documented on the `OpsImplementationConfig` field metadata and in
>   `docs/usage/arguments.md`.
> - `configs/dit/wan*.yaml` (3 files) — pin
>   `rotary_pos_emb_implementation: eager`.
>
> **Test infrastructure** — `tests/tools/training_utils.py` and downstream
>
> - `resolve_ops_overrides(model_name)` emits hardware-aware
>   `--model.ops_implementation.X=Y` flags for NPU torchruns;
>   `_NPU_PER_MODEL_OVERRIDES` covers DeepSeek-V3 (no NPU RMSNorm/RoPE) and
>   Qwen2-VL / Qwen2.5-VL / Qwen2.5-Omni (multimodal RoPE).
> - `make_eager_ops_config(**overrides)` and `make_npu_ops_config(model_name)`
>   replace `OpsImplementationConfig.all_eager()` across tests; both pin the
>   Qwen3.5 GatedDeltaNet trio to `eager` so tests don't pull in
>   `flash-linear-attention` on hosts that don't have it.
> - `model_name` threaded through `build_torchrun_cmd` /
>   `run_training_config` and into `tests/e2e/utils.py`,
>   `tests/distributed/test_fsdp_equivalence.py`, `tests/checkpoints/utils.py`,
>   `tests/data/{test_datasets,test_multisource_dataset,test_dynamic_batching_dataset}.py`.
> - `tests/{e2e/test_e2e_parallel,distributed/test_fsdp_equivalence}.py` —
>   skip `qwen3_5` / `qwen3_5_moe` parametrize cases on NPU (no NPU
>   GatedDeltaNet kernel; eager path only supports non-varlen, but the e2e
>   command uses `dyn_bsz=True`).
>
> **Docs** — `docs/usage/arguments.md`, `docs/design/kernel_selection.md`
>
> - Refresh defaults table, NPU validation contract, lifecycle "Ownership"
>   paragraph (now describes the strict-raise contract); replace
>   `VEOMNI_USE_LIGER_KERNEL` references in the transformers-v5 comparison
>   table with the corresponding `OpsImplementationConfig` field name.

### Test

> - `make quality` clean.
> - `pytest tests/ops/test_moe_hw_gate.py` — 15/15 pass.
> - `pytest tests/models/{test_padded_packed_loss,test_vlm_trainer}.py` —
>   pass; targeted `test_models_patch.py::test_models_patch_fwd_bwd[qwen2]`
>   passes.
> - `build_foundation_model(...)` with no `ops_implementation` raises
>   `ValueError` with the documented message; happy path with
>   `ops_implementation=make_eager_ops_config()` builds cleanly.
> - GPU host: constructed `OpsImplementationConfig()` with new defaults —
>   validates clean; logs show `liger_kernel` / `fused_triton` / `triton`
>   bound for every op.
> - Mocked `is_torch_npu_available()=True`: every GPU-only choice raises
>   with the listed alternatives. The all-NPU config (`fused_npu` +
>   `chunk_loss` + `npu` + `eager` + `triton`) passes.
> - Per-model strict-raise: simulated `apply_per_model_patches` for
>   Wan-shape `extra_backends` confirms `rotary_pos_emb=liger_kernel`
>   (default) raises with `"explicitly disabled for Wan"` +
>   `available=['eager','npu','triton']`; explicit
>   `rotary_pos_emb=eager` succeeds.
> - Verified Qwen3-VL `Qwen3VLTextRMSNorm` + `apply_rotary_pos_emb` are
>   HF-standard on transformers v4.57.3 and v5 — the `liger_kernel` lift
>   is safe.
>
> Pending CI:
>
> - `gpu_e2e_test.yml` / `gpu_unit_tests.yml` — should pass with the `gpu`
>   extra (ships liger-kernel + flash-attn).
> - `npu_e2e_test.yml` / `npu_unit_tests.yml` — `resolve_ops_overrides`
>   threaded through every torchrun in
>   `tests/{e2e,distributed,checkpoints,data}` emits NPU-supported
>   overrides per model.

### Checklist Before Submitting

- [x] Read the [Contribute Guide](https://github.com/ByteDance-Seed/VeOmni/blob/main/CONTRIBUTING.md)
- [x] Applied pre-commit checks
- [x] Added/updated documentation
- [x] Added tests to CI workflow (or explained why not feasible) —
  `resolve_ops_overrides` covers every model NPU CI exercises;
  `test_moe_hw_gate.py` pins all-eager. The change is config-default +
  validator behavior, exercised end-to-end by every
  `tests/e2e/test_e2e_parallel.py` parametrization.

### CODE REVIEW

**Comment by TimYangst** [comment]
@gemini-code-assist please help review the pr

File: veomni/arguments/arguments_types.py [hunk 1]
```diff
@@ -656,75 +673,76 @@ class OpsImplementationConfig:
...
+        """
+        return cls(
+            moe_implementation=overrides.pop("moe_implementation", "eager"),
+            cross_entropy_loss_implementation=overrides.pop("cross_entropy_loss_implementation", "eager"),
+            rms_norm_implementation=overrides.pop("rms_norm_implementation", "eager"),
+            swiglu_mlp_implementation=overrides.pop("swiglu_mlp_implementation", "eager"),
+            rotary_pos_emb_implementation=overrides.pop("rotary_pos_emb_implementation", "eager"),
+            load_balancing_loss_implementation=overrides.pop("load_balancing_loss_implementation", "eager"),
+            **overrides,  # any remaining (e.g. attn_implementation) flow through
+        )
```
[the remark this answers is not in the document]
Comment (TimYangst) [reply veomni/arguments/arguments_types.py]: Updated comments

File: veomni/arguments/arguments_types.py [hunk 2]
```diff
@@ -754,28 +772,40 @@ def __post_init__(self):
...
         """
-        from ..ops import config as ops_config_pkg  # noqa: F401  import side-effect
-        from ..ops.config.registry import list_ops
+        from ..utils.import_utils import is_package_available, is_torch_npu_available
 
-        for op in list_ops():
-            value = getattr(self, op.config_field)
-            backend = op.backends.get(value)
-            if backend is None:
+        on_npu = is_torch_npu_available()
```
[the remark this answers is not in the document]
Comment (TimYangst) [reply veomni/arguments/arguments_types.py:822]: Done

File: tests/models/test_models_patch.py [hunk 3]
```diff
@@ -17,6 +17,7 @@
     ModelArguments,
     TrainingArguments,
 )
+from veomni.arguments.arguments_types import OpsImplementationConfig
```
Comment (piyifan123) [inline tests/models/test_models_patch.py]: can we make it a default setting under tests instead of veomni?

File: tests/tools/training_utils.py [hunk 4]
```diff
@@ -16,9 +16,51 @@
...
-# Pick the fused-MoE backend that matches the test hardware. On NPU the NPU
-# kernel is the only option; on GPU default to Triton (SM70+).
-_FUSED_MOE_IMPL = "fused_npu" if is_torch_npu_available() else "fused_triton"
+# NPU ops_implementation overrides per model. The public
+# ``OpsImplementationConfig`` defaults are GPU-optimal (Liger / Triton) and
+# raise on NPU at config validation time, so every NPU test must override
+# every per-op field. ``_NPU_OPS_DEFAULTS`` is the baseline; entries in
+# ``_NPU_PER_MODEL_OVERRIDES`` (DeepSeek-V3, Qwen-VL family) pin specific
+# fields to ``eager`` where the model has no NPU kernel.
+_NPU_OPS_DEFAULTS: Dict[str, str] = {
```
Comment (piyifan123) [inline tests/tools/training_utils.py:26]: can we just use this in NPU tests instead of all eager?

**Review by piyifan123 — APPROVED** [review]

File: tests/models/test_models_patch.py (hunk 3 above)
Comment (TimYangst) [reply tests/models/test_models_patch.py]: Done

File: tests/tools/training_utils.py (hunk 4 above)
Comment (TimYangst) [reply tests/tools/training_utils.py:26]: Done

**Comment by TimYangst** [comment]
@gemini-code-assist please help review the PR

File: veomni/arguments/arguments_types.py [hunk 5]
```diff
@@ -625,24 +624,56 @@ def _resolve_profile(self):
...
+    "swiglu_mlp_implementation": frozenset(),
+    "load_balancing_loss_implementation": frozenset({"triton"}),
+    "cross_entropy_loss_implementation": frozenset({"chunk_loss", "npu"}),
+    "moe_implementation": frozenset({"fused_npu"}),
+}
+
+_NPU_REQUIRED: Dict[str, frozenset] = {
+    "cross_entropy_loss_implementation": frozenset({"npu"}),
+    "moe_implementation": frozenset({"fused_npu"}),
+}
```
[the remark this answers is not in the document]
Comment (TimYangst) [reply veomni/arguments/arguments_types.py:649]: Done

File: veomni/arguments/arguments_types.py [hunk 6]
```diff
@@ -625,24 +624,56 @@ def _resolve_profile(self):
...
+# triton-ascend for load-balancing loss vs. CUDA-only mainline triton for
+# DeepSeek-V3 RMSNorm / RoPE.
+_NPU_ALLOWED: Dict[str, frozenset] = {
+    "rms_norm_implementation": frozenset({"npu"}),
+    "rotary_pos_emb_implementation": frozenset({"npu"}),
+    "swiglu_mlp_implementation": frozenset(),
+    "load_balancing_loss_implementation": frozenset({"triton"}),
+    "cross_entropy_loss_implementation": frozenset({"chunk_loss", "npu"}),
+    "moe_implementation": frozenset({"fused_npu"}),
+}
```
[the remark this answers is not in the document]
Comment (TimYangst) [reply veomni/arguments/arguments_types.py:642]: won't fix for now

**Review by Coach257 — APPROVED** [review]
